### PR TITLE
[Testing] add account balance coverage to account settlement

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Add coverage for the following balances to the account settlement tests:
- Application/supplier account
- Application/supplier module account
- Application/supplier stake

## Issue

- #584

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

## Summary by CodeRabbit

- **New Features**
	- Introduced new `SupplierKeeper` to manage supplier interactions within the tokenomics module.

- **Tests**
	- Enhanced test cases to include new balance checks and validations.
	- Added new function `getBalance` for retrieving account balances in tests.
	- Updated test logic to handle various accounting scenarios, including debt and low app stake.

### CODE REVIEW

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 1]
```diff
@@ -57,24 +63,290 @@ func TestSettleSessionAccounting_HandleAppGoingIntoDebt(t *testing.T) {
...
 	require.NoError(t, err)
-	// TODO_TEST: Need to make sure the application is unstaked at this point in time.
 }
 
 func TestSettleSessionAccounting_ValidAccounting(t *testing.T) {
-	t.Skip("TODO_BLOCKER(@Olshansk): Add E2E and integration tests so we validate the actual state changes of the bank & account keepers.")
-	// Assert that `suppliertypes.ModuleName` account module balance is *unchanged*
-	// Assert that `supplierAddress` account balance has *increased* by the appropriate amount
-	// Assert that `supplierAddress` staked balance is *unchanged*
-	// Assert that `apptypes.ModuleName` account module balance is *unchanged*
```
Comment (bryanchriswhite) [inline x/tokenomics/keeper/settle_session_accounting_test.go:68]: @reviewer, this assertion seems to be incorrect to me. My understanding is that the account module balance MUST *decrease* as a result of the burn. I assumed this is copy-pasta and have the test making the modified assertion.

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 2]
```diff
@@ -57,24 +64,188 @@ func TestSettleSessionAccounting_HandleAppGoingIntoDebt(t *testing.T) {
...
+
+	// Set compute_units_to_tokens_multiplier to 1 to simplify expectation calculations.
+	err := keepers.Keeper.SetParams(ctx, tokenomicstypes.Params{
+		ComputeUnitsToTokensMultiplier: 1,
+	})
+	require.NoError(t, err)
+
+	// Add a new application
+	appStake := cosmostypes.NewCoin("upokt", math.NewInt(40000))
+	expectedAppEndStakeAmount := cosmostypes.NewCoin("upokt", math.NewInt(0))
```
Comment (red-0ne) [inline x/tokenomics/keeper/settle_session_accounting_test.go:172]:
```suggestion
	expectedAppEndStakeZeroAmount := cosmostypes.NewCoin("upokt", math.NewInt(0))
```

**Review by red-0ne — COMMENTED** [review]
Left a few comments but this reflects well how it works.

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 3]
```diff
@@ -57,24 +64,188 @@ func TestSettleSessionAccounting_HandleAppGoingIntoDebt(t *testing.T) {
...
+
+	// Set compute_units_to_tokens_multiplier to 1 to simplify expectation calculations.
+	err := keepers.Keeper.SetParams(ctx, tokenomicstypes.Params{
+		ComputeUnitsToTokensMultiplier: 1,
+	})
+	require.NoError(t, err)
+
+	// Add a new application
+	appStake := cosmostypes.NewCoin("upokt", math.NewInt(1000000))
+	expectedAppEndStakeAmount := cosmostypes.NewCoin("upokt", math.NewInt(420))
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/tokenomics/keeper/settle_session_accounting_test.go:82]: This is just an arbitrary amount which the `expectedAppBurn` is defined in terms of. The idea is to ensure a non-zero stake balance that can be asserted at the end of the test. Defining this value first lets us define `expectedAppBurn` in terms of it and `appStake`, which I figured would be the simplest way to set things up.

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 4]
```diff
@@ -57,24 +64,188 @@ func TestSettleSessionAccounting_HandleAppGoingIntoDebt(t *testing.T) {
...
 	// Assert that `applicationAddress` account balance is *unchanged*
+	appEndBalance := getBalance(t, ctx, keepers, app.GetAddress())
+	require.EqualValues(t, appStartBalance, appEndBalance)
+
 	// Assert that `applicationAddress` staked balance has decreased by the appropriate amount
+	app, appIsFound := keepers.GetApplication(ctx, app.GetAddress())
+	require.True(t, appIsFound)
+	require.Equal(t, &expectedAppEndStakeAmount, app.GetStake())
+
+	// Assert that `apptypes.ModuleName` account module balance is *decreased* by the appropriate amount
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/tokenomics/keeper/settle_session_accounting_test.go:138]:
```
	// Assert that `apptypes.ModuleName` account module balance is *decreased* by the appropriate amount
	// NB: The application module account burns the amount of uPOKT that was held in escrow
	// on behalf of the applications which were serviced in a given session.
```
```
	// Assert that `suppliertypes.ModuleName` account module balance is *unchanged*
	// NB: Supplier rewards are minted to the supplier module account but then immediately
	// distributed to the supplier accounts which provided service in a given session.
```

**Pull request merged**